### ISSUE TITLE
High Latency Mitigation allowing for impossible double weaves

### ISSUE THREAD

**Issue by SpoonOfTate** [Soreepeong/XivAlexander#140]
High Latency Mitigation Mode 3 allows for actions to be taken that are otherwise impossible, such as double weaving during the window in which a movement ability was used.

![image](https://user-images.githubusercontent.com/89931028/131715532-424f845e-d963-4278-8c17-1873cb979dd6.png)

Red Mage at 2.37 GCD performing Corps-a-corps and Displacement, then successfully casting Veraero with no clipping;
![displacement double weave](https://user-images.githubusercontent.com/89931028/131716331-f8632e50-e2dd-46e3-be5b-3bc9cad64636.gif)

Dragoon at 2.44 GCD double weaving a Jump/Dive with another oGCD, which is otherwise impossible due to Jump/Dive animation lock.
![dragoon jump double weave 4](https://user-images.githubusercontent.com/89931028/131717196-eb16ab60-5255-4382-9ae6-6e031d881bec.gif)

Due to the fact that this is possibly time travel, this seems to be a very pressing issue. This behavior is not present at Mode 1.

**Comment by Spanglish1337** [comment]
Can you still double weave normally with mode 1? Mode 3 was selected as default when I installed the plugin, I don't play RDM nor Dragoon yet.

**Comment by Datachange** [comment]
Can we please get more information about the logic behind how this works?
Many abilities don't have a 0.5s animation lock. Some have more some (may) have less.
Is there logic based on which ability is used on how it changes the client's cooldown timer?

**Comment by Soreepeong** [comment]
https://github.com/Soreepeong/XivAlexander/blob/709a8f9c615be03781cbe82b14e34c9508e9a45c/XivAlexander/App_Feature_AnimationLockLatencyHandler.cpp#L306-L386

All the work happens in here - you can try figuring out what's going wrong yourself, as it works fine for me, as in actions will go out just like how it feels like to play on KR servers without XivAlex on (#25). One thing to note is that if your network does not allow pinging SE servers, the program will fall back to resolving latency from socket ioctl, which tends to be a little bit higher than the latency reported from `ping` command. 

By the way, I've just tested again with the latest release, and it does work fine for me. Attach logs from XivAlex to let me see what went wrong. That 500ms thing looks like you're on a really high ping...

```
## Mana
12.959	Request: actionId=08ce sequence=0003 delay=600+15963ms prevNextRelative=10000ms+
13.059	Effect: actionId=08ce sourceSequence=0003 rtt=93ms socketLatency=33ms latencyEstimate=52ms delayAdjusted=41ms wait=600ms->548ms next=07:31:13.6070901
13.593	Request: actionId=08c1 sequence=0004 delay=600+25ms prevNextRelative=-16ms
13.660	Effect: actionId=08c1 sourceSequence=0004 rtt=62ms socketLatency=33ms latencyEstimate=43ms delayAdjusted=19ms wait=600ms->573ms next=07:31:14.2335747
14.228	Request: actionId=1d75 sequence=0005 delay=600+40ms prevNextRelative=5ms
14.328	Effect: actionId=1d75 sourceSequence=0005 rtt=94ms socketLatency=33ms latencyEstimate=54ms delayAdjusted=40ms wait=600ms->546ms next=07:31:14.8745654
14.864	Request: actionId=1d76 sequence=0006 delay=600+41ms prevNextRelative=1ms
14.948	Effect: actionId=1d76 sourceSequence=0006 rtt=78ms socketLatency=33ms latencyEstimate=50ms delayAdjusted=28ms wait=600ms->550ms next=07:31:15.4986075
15.481	Request: actionId=4068 sequence=0007 delay=600+9ms prevNextRelative=-19ms
15.598	Effect: actionId=4068 sourceSequence=0007 rtt=110ms socketLatency=33ms latencyEstimate=57ms delayAdjusted=46ms wait=600ms->555ms next=07:31:16.1534901
16.183	Request: actionId=08d8 sequence=0008 delay=600+103ms prevNextRelative=38ms
16.249	Effect: actionId=08d8 sourceSequence=0008 rtt=63ms socketLatency=33ms latencyEstimate=43ms delayAdjusted=20ms wait=600ms->557ms next=07:31:16.8068233
16.852	Request: actionId=08c5 sequence=0009 delay=600+72ms prevNextRelative=52ms
16.900	Effect: actionId=08c5 sourceSequence=0009 rtt=47ms socketLatency=33ms latencyEstimate=37ms delayAdjusted=19ms wait=600ms->572ms next=07:31:17.4728905

## Primal
53.531	Request: actionId=08ce sequence=0001 delay=0+359300187ms prevNextRelative=10000ms+
53.764	Effect: actionId=08ce sourceSequence=0001 rtt=234ms socketLatency=142ms latencyEstimate=173ms delayAdjusted=61ms wait=600ms->427ms next=07:33:54.1912699
54.182	Request: actionId=08c1 sequence=0002 delay=600+56ms prevNextRelative=-5ms
54.399	Effect: actionId=08c1 sourceSequence=0002 rtt=219ms socketLatency=142ms latencyEstimate=170ms delayAdjusted=49ms wait=600ms->435ms next=07:33:54.8340242
54.833	Request: actionId=08c5 sequence=0003 delay=600+57ms prevNextRelative=3ms
55.084	Effect: actionId=08c5 sourceSequence=0003 rtt=250ms socketLatency=142ms latencyEstimate=182ms delayAdjusted=68ms wait=600ms->418ms next=07:33:55.5020505
55.502	Request: actionId=1d75 sequence=0004 delay=600+56ms prevNextRelative=-12ms
55.752	Effect: actionId=1d75 sourceSequence=0004 rtt=250ms socketLatency=142ms latencyEstimate=183ms delayAdjusted=67ms wait=600ms->429ms next=07:33:56.1813243
56.354	Request: actionId=08d8 sequence=0005 delay=600+259ms prevNextRelative=180ms
56.620	Effect: actionId=08d8 sourceSequence=0005 rtt=266ms socketLatency=142ms latencyEstimate=187ms delayAdjusted=73ms wait=600ms->407ms next=07:33:57.0274623
57.022	Request: actionId=1d76 sequence=0006 delay=600+72ms prevNextRelative=-1ms
57.322	Effect: actionId=1d76 sourceSequence=0006 rtt=297ms socketLatency=142ms latencyEstimate=190ms delayAdjusted=83ms wait=600ms->387ms next=07:33:57.7096066
```

**Comment by SpoonOfTate** [comment]
I actually raised a similar issue in another plugin, NoClippy, and while it looked like I was doing something impossible, I wasn't actually taking into account very small clips that happened within less than ~150 ms. Funnily enough I just saw you commented on the same issue! A huge part of my confusion was definitely my own inexperience with both the jobs and with low latency in general; I live in SEA and play on the NA datacenters, so really low latency felt really suspicious simply because it felt too good to be true. 

Thank you so much for putting in the work I couldn't and subsequently easing my anxieties.
